File: count-uniques/count-uniques.py
#!/usr/bin/env python3

import argparse
import re
import sys
from itertools import combinations


def split_row(line: str) -> list[str]:
    if "\t" in line:
        return [cell.strip() for cell in line.split("\t")]
    return re.split(r"\s+", line.strip())


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Count first-seen value pairs per test-case row."
    )
    parser.add_argument(
        "input_file",
        nargs="?",
        default="-",
        help="Path to PICT output file (or - / omitted for stdin)",
    )
    parser.add_argument(
        "--no-header",
        action="store_true",
        help="Treat the first non-empty row as data (do not skip header)",
    )
    parser.add_argument(
        "--reorder",
        action="store_true",
        help="Output test rows reordered by first-seen pair count (descending)",
    )
    args = parser.parse_args()

    seen_pairs: set[tuple[tuple[int, str], tuple[int, str]]] = set()
    first_data_row_seen = False
    header_line = ""
    scored_rows: list[tuple[int, str]] = []

    handle = (
        sys.stdin
        if args.input_file == "-"
        else open(args.input_file, "r", encoding="utf-8")
    )
    with handle:
        for raw_line in handle:
            raw = raw_line.rstrip("\n")
            line = raw.strip()
            if not line:
                continue

            if not args.no_header and not first_data_row_seen:
                header_line = raw
                first_data_row_seen = True
                continue

            first_data_row_seen = True
            values = split_row(line)
            new_pairs = 0

            for left, right in combinations(range(len(values)), 2):
                pair = ((left, values[left]), (right, values[right]))
                if pair not in seen_pairs:
                    seen_pairs.add(pair)
                    new_pairs += 1

            if args.reorder:
                scored_rows.append((new_pairs, raw))
            else:
                print(f"{raw}\t# [{new_pairs}]")

    if args.reorder:
        if header_line:
            print(header_line)
        for _, row in sorted(scored_rows, key=lambda item: item[0], reverse=True):
            print(row)


if __name__ == "__main__":
    main()

File: scripts/count-uniques.pl
#!/usr/bin/env perl

use strict;
use warnings;
use Getopt::Long qw(GetOptions);

binmode STDIN,  ':encoding(UTF-8)';
binmode STDOUT, ':encoding(UTF-8)';
binmode STDERR, ':encoding(UTF-8)';

my $no_header = 0;
my $reorder = 0;

GetOptions(
    'no-header' => \$no_header,
    'reorder'   => \$reorder,
) or die "Usage: perl count-uniques.pl [--no-header] [--reorder] [input_file]\n";

my $input_file = shift @ARGV;
$input_file = '-' unless defined $input_file;

die "Usage: perl count-uniques.pl [--no-header] [--reorder] [input_file]\n" if @ARGV;

sub split_row {
    my ($line) = @_;

    if (index($line, "\t") >= 0) {
        my @parts = split /\t/, $line, -1;
        s/^\s+|\s+$//g for @parts;
        return @parts;
    }

    return split /\s+/, $line;
}

my %seen_pairs;
my $first_data_row_seen = 0;
my $header_line = q{};
my @scored_rows;
my $row_index = 0;

my $handle;
if ($input_file eq '-') {
    $handle = *STDIN;
}
else {
    open my $file_handle, '<:encoding(UTF-8)', $input_file
      or die "Cannot open $input_file: $!\n";
    $handle = $file_handle;
}

while (my $raw_line = <$handle>) {
    chomp $raw_line;
    my $line = $raw_line;
    $line =~ s/^\s+|\s+$//g;

    next if $line eq q{};

    if (!$no_header && !$first_data_row_seen) {
        $header_line = $raw_line;
        $first_data_row_seen = 1;
        next;
    }

    $first_data_row_seen = 1;

    my @values = split_row($line);
    my $new_pairs = 0;

    for (my $left = 0; $left < @values; ++$left) {
        for (my $right = $left + 1; $right < @values; ++$right) {
            my $pair_key = join "\x1F",
              $left, $values[$left], $right, $values[$right];
            if (!exists $seen_pairs{$pair_key}) {
                $seen_pairs{$pair_key} = 1;
                ++$new_pairs;
            }
        }
    }

    if ($reorder) {
        push @scored_rows, [$new_pairs, $row_index++, $raw_line];
    }
    else {
        print "$raw_line\t# [$new_pairs]\n";
    }
}

if ($reorder) {
    print "$header_line\n" if $header_line ne q{};
    for my $entry (sort { $b->[0] <=> $a->[0] || $a->[1] <=> $b->[1] } @scored_rows) {
        print "$entry->[2]\n";
    }
}
